### PR TITLE
DM-11595: Use mkdtemp for test directories

### PR DESCRIPTION
**Pull request by timj**
@n8pease work in progress. In particular test_repository.TestBasics doesn't work and I don't know why...

### CODE REVIEW

File: tests/test_mapperImport.py [hunk 1]
```diff
@@ -38,15 +39,16 @@ class MapperImportTestCase(unittest.TestCase):
     """A test case for the data butler finding a Mapper in a root"""
 
     def setUp(self):
-        if os.path.exists(os.path.join(ROOT, 'root/out')):
-            shutil.rmtree(os.path.join(ROOT, 'root/out'))
+        self.testDir = os.path.join(ROOT, 'root')
```
Comment (n8pease) [inline tests/test_mapperImport.py]: nit: I think inputDir would be more accurate?

File: tests/test_mapperImport.py [hunk 2]
```diff
@@ -38,15 +39,16 @@ class MapperImportTestCase(unittest.TestCase):
...
+        self.rootPath = tempfile.mkdtemp(dir=self.testDir, prefix="out-")
+        self.outPath = os.path.split(self.rootPath)[1]
```
Comment (n8pease) [inline tests/test_mapperImport.py]: by removing the absolute part of the path and passing only the temp dir name to the outPath iarg, you're creating an output repository relative to the current working dir. Which would be fine, but in tearDown you're deleting self.rootPath, not self.outPath, so outpath will never get cleaned up. Right? I think you can just pass self.rootPath to the outPath iarg. 
Or maybe I'm faded from travel. I need to board... off I go.

File: tests/test_mapperImport.py [hunk 3]
```diff
@@ -56,8 +58,8 @@ def checkIO(self, butler, bbox, ccd):
         butler.put(bbox, "x", ccd=ccd)
         y = butler.get("x", ccd=ccd, immediate=True)
         self.assertEqual(bbox, y)
-        self.assert_(os.path.exists(
-            os.path.join(ROOT, "root", "out", "foo%d.pickle" % ccd)))
+        self.assertTrue(os.path.exists(
+            os.path.join(self.rootPath, "foo%d.pickle" % ccd)))
```
Comment (n8pease) [inline tests/test_mapperImport.py]: I think this will fail if the current working dir is not the same as the directory above outpath?

Comment (n8pease) [inline tests/test_mapperImport.py]: confirmed; running tests from inside `daf_persistence/python` (as an example) fails
```
n8pease@Nates-MacBook-Pro ~/3/lsstsw/build/daf_persistence/python[tickets/DM-11595*]$ py.test ~/3/lsstsw/build/daf_persistence/tests/test_mapperImport.py 
```

File: tests/test_reposInButler.py [hunk 4]
```diff
@@ -98,15 +99,12 @@ def map_str(self, dataId, write):
...
 class ReposInButler(unittest.TestCase):
 
-    def clean(self):
-        if os.path.exists(os.path.join(ROOT, 'repoOfRepos')):
-            shutil.rmtree(os.path.join(ROOT, 'repoOfRepos'))
-
-    def setup(self):
-        self.clean()
+    def setUp(self):
+        self.testDir = tempfile.mkdtemp(dir=ROOT, prefix='repoOfRepos-')
```
Comment (n8pease) [inline tests/test_reposInButler.py]: not your bug but while your here you could fix "repoOfRepos" to accurately show the test class name "reposInButler"

File: tests/test_repository.py [hunk 5]
```diff
@@ -374,10 +378,12 @@ class TestMultipleInputs(unittest.TestCase):
         - verify dissimilar datasets with same id that were written to the repos
 
     """
+    def setUp(self):
+        self.testDir = tempfile.mkdtemp(dir=ROOT, prefix="TestMapperInference-")
```
Comment (n8pease) [inline tests/test_repository.py]: the prefix here should be "TestMultipleInputs"

File: tests/test_repository.py [hunk 6]
```diff
@@ -763,9 +768,8 @@ class TestParentRepository(unittest.TestCase):
     """A test to verify that when a parent repository is used the correct one is found."""
 
     def setUp(self):
-        self.testDir = os.path.join(ROOT, 'TestParentRepository')
-        if os.path.exists(self.testDir):
-            shutil.rmtree(self.testDir)
+        self.testDir = tempfile.mkdtemp(dir=ROOT, prefix="TestBasics-")
```
Comment (n8pease) [inline tests/test_repository.py]: prefix should be "TestParentRepository"

File: tests/test_repositoryCfg.py [hunk 7]
```diff
@@ -39,102 +39,105 @@
 class TestCfgRelationship(unittest.TestCase):
 
     def setUp(self):
-        self.tearDown()
+        self.testDir = tempfile.mkdtemp(dir=ROOT,
+                                        prefix="repositoryCfg-")
```
Comment (n8pease) [inline tests/test_repositoryCfg.py]: prefix should be TestCfgRelationship

File: tests/test_repositoryCfg.py [hunk 8]
```diff
@@ -157,37 +160,36 @@ def testStorageRepoCfgCache(self):
...
+        self.assertEqual(cfg, storage.repositoryCfgs[os.path.join(self.testDir, 'a')])
 
 class TestNestedCfg(unittest.TestCase):
 
-    rootDir = os.path.join(ROOT, 'repositoryCfg_TestNestedCfg')
-
     def setUp(self):
-        self.tearDown()
+        self.testDir = tempfile.mkdtemp(dir=ROOT,
+                                        prefix="repositoryCfg_TestNestedCfg-")
```
Comment (n8pease) [inline tests/test_repositoryCfg.py]: I see you just used what was already there but the prefix here can probably just be TestNestedCfg

File: tests/test_repositoryCfg.py [hunk 9]
```diff
@@ -214,19 +216,18 @@ class TestCfgFileVersion(unittest.TestCase):
     """
 
     def setUp(self):
-        self.tearDown()
+        self.testDir = tempfile.mkdtemp(dir=ROOT,
+                                        prefix="repositoryCfg-")
```
Comment (n8pease) [inline tests/test_repositoryCfg.py]: prefix -> TestCfgFileVersion

File: tests/test_safeFileIo.py [hunk 10]
```diff
@@ -45,37 +46,37 @@ def setup_module(module):
 class WriteOnceCompareSameTest(unittest.TestCase):
 
     def setUp(self):
-        self.tearDown()
+        self.testDir = tempfile.mkdtemp(dir=ROOT, prefix='safeFileIo-')
```
Comment (n8pease) [inline tests/test_safeFileIo.py]: prefix -> "WriteOnceCompareSameTest"

File: tests/test_repository.py [hunk 11]
```diff
@@ -763,9 +768,8 @@ class TestParentRepository(unittest.TestCase):
...
+        self.testDir = os.path.join(self.testDir, 'TestParentRepository')
```
Comment (n8pease) [inline tests/test_repository.py]: this line shouldn't be here? It's preventing files inside `TestBasics-...` (which should be `TestParentRepository-...`) 
from getting removed.

Comment (timj) [inline tests/test_repository.py]: Of course. Thanks for spotting that. I had to change many many lines...

**Pull request merged**